File: src/main/cpp/solver/IdentifyPairs.cpp
#include "IdentifyPairs.h"
#include "SudokuIndexes.h"

using namespace std;

SudokuSolverStatus IdentifyPairs::run(SudokuResultShared result)
{
    SudokuPlayerBoard &board = *result->getPlayerBoard();
    for (int index = 0; index < 81; index++)
    {
        if (!board.isEmptyCell(index))
        {
            continue;
        }

        SudokuValueSet allowedValues = board.getAllowedValues(index);
        if (allowedValues.size() != 2)
        {
            continue;
        }

        int peerIndex = findPeer(board, allowedValues, getRelated(index));
        if (peerIndex < 0)
        {
            continue;
        }

        // found peer
        int eliminationCount = 0;
        int row = rowFromIndex(index);
        if (row == rowFromIndex(peerIndex))
        {
            // same row
            if (tryEliminate(board, index, peerIndex, allowedValues, getRowIndexes(row)))
            {
                eliminationCount++;
            }
        }

        int col = columnFromIndex(index);
        if (col == columnFromIndex(peerIndex))
        {
            // same column
            if (tryEliminate(board, index, peerIndex, allowedValues, getColumnIndexes(col)))
            {
                eliminationCount++;
            }
        }

        int square = squareFromIndex(index);
        if (square == squareFromIndex(peerIndex))
        {
            // same square
            if (tryEliminate(board, index, peerIndex, allowedValues, getSquareIndexes(square)))
            {
                eliminationCount++;
            }
        }

        if (eliminationCount > 0)
        {
            result->report(SudokuStepComplexity::HARD, SudokuStep::IDENTIFY_PAIRS);
            return SudokuSolverStatus::SUCCEEDED;
        }
    }

    return SudokuSolverStatus::UNKNOWN;
}

bool IdentifyPairs::tryEliminate(
    SudokuPlayerBoard &board, int ignore1, int ignore2, SudokuValueSet allowedValues, const sequence &indexes)
{
    bool found = false;
    for (int temp = 0; temp < 9; temp++)
    {
        int index = indexes.at(temp);
        if (index == ignore1 || index == ignore2)
        {
            continue;
        }

        if (!board.isEmptyCell(index))
        {
            continue;
        }

        SudokuValueSet tempAllowedValues = board.getAllowedValues(index);
        if ((tempAllowedValues & allowedValues).size() > 0)
        {
            // found a cell that we can remove values - turn them off
            board.disallowValues(index, allowedValues);
            found = true;
        }
    }

    return found;
}

int IdentifyPairs::findPeer(SudokuPlayerBoard &board, SudokuValueSet allowedValues, const related &indexes)
{
    for (int peerIndex : indexes)
    {
        if (!board.isEmptyCell(peerIndex))
        {
            continue;
        }
        SudokuValueSet peerAllowedValues = board.getAllowedValues(peerIndex);
        if (peerAllowedValues == allowedValues)
        {
            return peerIndex;
        }
    }
    return -1;
}

File: go/solver/identify_pairs.go
package solver

import (
	"context"

	"github.com/nissimnatanov/des/go/boards"
	"github.com/nissimnatanov/des/go/boards/indexes"
	"github.com/nissimnatanov/des/go/boards/values"
)

type identifyPairs struct {
}

func (a identifyPairs) Run(ctx context.Context, state AlgorithmState) Status {
	b := state.Board()
	// force local alloc for cleaner profile
	var peersCache [5]int
	var eliminationCount int
	freeCellOnStart := b.FreeCellCount()

	for index, allowed := range b.AllAllowedValues {
		if allowed.Size() != 2 {
			// this includes non-empty cells too (allowed set is empty)
			continue
		}
		peers := a.findPeers(b, index, allowed, indexes.RelatedSequence(index), peersCache[:])
		if len(peers) == 0 {
			// no peer found
			continue
		}

		status := a.tryEliminate(state, index, peers, allowed, indexes.RowFromIndex, indexes.RowSequence)
		if status == StatusSucceeded {
			eliminationCount++
			// do not stop yet if we found a solution, let's check other peers
		} else if status != StatusUnknown {
			return status
		}

		status = a.tryEliminate(state, index, peers, allowed, indexes.ColumnFromIndex, indexes.ColumnSequence)
		if status == StatusSucceeded {
			eliminationCount++
			// do not stop yet if we found a solution, let's check other peers
		} else if status != StatusUnknown {
			return status
		}

		status = a.tryEliminate(state, index, peers, allowed, indexes.SquareFromIndex, indexes.SquareSequence)
		if status == StatusSucceeded {
			eliminationCount++
			// do not stop yet if we found a solution, let's check other peers
		} else if status != StatusUnknown {
			return status
		}
		if b.FreeCellCount() < freeCellOnStart {
			// if we set at least one value, we can stop now and try faster algos
			return StatusSucceeded
		}
	}
	if eliminationCount > 0 {
		// if we found at least one index that lead to elimination, let's stop
		// and go back to cheaper algorithm such as theOnlyChoice
		return StatusSucceeded
	}
	return StatusUnknown
}

func (a identifyPairs) tryEliminate(
	state AlgorithmState, index int, peers []int,
	allowed values.Set,
	seqNumFromIndex func(int) int,
	indexesFromSeq func(int) indexes.Sequence,
) Status {
	seqNum := seqNumFromIndex(index)
	var seqPeer = -1

	for _, peer := range peers {
		if seqNum != seqNumFromIndex(peer) {
			continue
		}
		if seqPeer != -1 {
			// we already found one peer in the same sequence, this is second
			// in the same sequence which means there are
			// 3 cells with same pair of values
			state.AddStep(Step(a.String()), a.Complexity(), 1)
			return StatusNoSolution
		}
		seqPeer = peer
	}
	if seqPeer == -1 {
		// we didn't find at least one peer in the same sequence
		return StatusUnknown
	}

	return a.tryEliminateSeq(state, index, seqPeer, allowed, indexesFromSeq(seqNum))
}

func (a identifyPairs) tryEliminateSeq(
	state AlgorithmState, p1, p2 int,
	toEliminate values.Set, seq indexes.Sequence,
) Status {
	status := StatusUnknown
	board := state.Board()
	for _, index := range seq {
		if index == p1 || index == p2 || !board.IsEmpty(index) {
			continue
		}

		tempAllowed := board.AllowedValues(index)
		if values.Intersect(tempAllowed, toEliminate).Size() == 0 {
			// no intersection, continue
			continue
		}

		// found a cell that we can remove values - turn them off
		state.AddStep(Step(a.String()), a.Complexity(), 1)
		status = StatusSucceeded

		// since we are here, we can now easily check if we have only one allowed value left
		tempAllowed = tempAllowed.Without(toEliminate)
		switch tempAllowed.Size() {
		case 0:
			// no allowed values left, this is a dead end
			return StatusNoSolution
		case 1:
			// only one allowed value left, let's set it
			board.Set(index, tempAllowed.First())
		default:
			board.DisallowValues(index, toEliminate)
		}
	}

	return status
}

func (a identifyPairs) findPeers(
	board *boards.Game,
	index int,
	allowed values.Set,
	related indexes.Sequence,
	peersCache []int,
) []int {
	peers := peersCache[:0]
	for _, peerIndex := range related {
		if peerIndex < index {
			// we only need to search forward since we already searched for peers
			// of the previous indexes
			continue
		}
		if !board.IsEmpty(peerIndex) {
			continue
		}
		peerAllowed := board.AllowedValues(peerIndex)
		if peerAllowed == allowed {
			peers = append(peers, peerIndex)
		}
		// keep going to find more peers, this helps us to invalidate boards if > one peer is found
	}
	return peers
}

func (a identifyPairs) Complexity() StepComplexity {
	return StepComplexityHard
}

func (a identifyPairs) String() string {
	return "Identify Pairs"
}
